File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/render/wobble/WobbleOutlineBuilder.kt
package app.pbbls.android.features.path.render.wobble

import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.sin

/**
 * The wobbled render of one artwork piece, as pure geometry (no Compose or
 * `android.graphics` types, so the whole pipeline is JVM unit-testable —
 * same rule as `PebbleSvgModel`). `WobbleShapes.kt` converts it to a Compose
 * [androidx.compose.ui.graphics.Path] at the view layer.
 *
 * [ink] is the leaky filled outline: closed contours, filled together under
 * the nonzero winding rule. [centerline] is the displaced centerline the
 * future appear animation's reveal mask must stroke along (fat trimmed
 * stroke, mirroring iOS `PebbleAnimatedRenderView` — not wired up yet;
 * Android has no draw-on animation).
 */
internal class WobbleArt(
    val ink: List<List<WobblePoint>>,
    val centerline: List<WobblePolyline>,
)

/**
 * Converts flattened centerline polylines into wobbled filled outlines,
 * porting iOS `WobbleOutlineBuilder.swift` (itself the playground's
 * baked-export algorithm): offset both stroke edges from the centerline
 * first, then displace every contour point independently. The independent
 * displacement is what makes the width breathe ("leaky") — a stroked wobbled
 * centerline would stay constant-width.
 */
internal object WobbleOutlineBuilder {
    /**
     * End caps are semicircles of [CAP_SEGMENTS] arcs (CAP_SEGMENTS − 1
     * interior points), matching the playground.
     */
    private const val CAP_SEGMENTS = 6

    /**
     * A zero-length polyline (single-tap carve dot) becomes a circle with
     * this many segments. The playground drops dots; we keep them visible,
     * matching the current renderer's round-cap dot.
     */
    private const val DOT_SEGMENTS = 12

    fun art(
        polylines: List<WobblePolyline>,
        halfWidth: Double,
        params: WobbleParams,
        noise: SVGTurbulence,
    ): WobbleArt {
        val ink = mutableListOf<List<WobblePoint>>()
        val centerline = mutableListOf<WobblePolyline>()
        for (polyline in polylines) {
            for (contour in contours(polyline, halfWidth)) {
                if (contour.size <= 2) continue
                ink.add(contour.map { params.displace(it, noise) })
            }
            centerline.add(displacedCenterline(polyline, params, noise))
        }
        return WobbleArt(ink, centerline)
    }

    /**
     * Un-displaced outline contours for one polyline. Internal so tests can
     * assert point counts and winding without decoding a displaced path.
     */
    fun contours(
        polyline: WobblePolyline,
        halfWidth: Double,
    ): List<List<WobblePoint>> {
        val points = polyline.points
        val count = points.size

        if (count == 1) {
            val cx = points[0].x
            val cy = points[0].y
            val circle =
                (0 until DOT_SEGMENTS).map { i ->
                    val angle = 2 * Math.PI * i / DOT_SEGMENTS
                    WobblePoint(cx + halfWidth * cos(angle), cy + halfWidth * sin(angle))
                }
            return listOf(circle)
        }

        val normals = normals(points, polyline.isClosed)
        val left =
            (0 until count).map { i ->
                WobblePoint(points[i].x + normals[i].x * halfWidth, points[i].y + normals[i].y * halfWidth)
            }
        val right =
            (0 until count).map { i ->
                WobblePoint(points[i].x - normals[i].x * halfWidth, points[i].y - normals[i].y * halfWidth)
            }

        if (polyline.isClosed) {
            // Outer ring + reversed inner ring: opposite windings make the
            // pair render as an annulus under nonzero fill.
            return listOf(left, right.reversed())
        }

        val contour = left.toMutableList()
        appendCapArc(
            contour = contour,
            center = points[count - 1],
            from = left[count - 1],
            halfWidth = halfWidth,
            tangentX = points[count - 1].x - points[count - 2].x,
            tangentY = points[count - 1].y - points[count - 2].y,
        )
        contour.addAll(right.reversed())
        appendCapArc(
            contour = contour,
            center = points[0],
            from = right[0],
            halfWidth = halfWidth,
            tangentX = points[0].x - points[1].x,
            tangentY = points[0].y - points[1].y,
        )
        return listOf(contour)
    }

    // ── Pieces ─────────────────────────────────────────────────

    /**
     * Per-point normals from neighbor tangents; endpoints use their single
     * adjacent segment, closed rings wrap cyclically.
     */
    private fun normals(
        points: List<WobblePoint>,
        closed: Boolean,
    ): List<WobblePoint> {
        val count = points.size
        return (0 until count).map { i ->
            val prev = if (closed) points[(i - 1 + count) % count] else points[maxOf(0, i - 1)]
            val next = if (closed) points[(i + 1) % count] else points[minOf(count - 1, i + 1)]
            val tx = next.x - prev.x
            val ty = next.y - prev.y
            val rawLength = hypot(tx, ty)
            val length = if (rawLength == 0.0) 1.0 else rawLength // playground: `|| 1`
            WobblePoint(-ty / length, tx / length)
        }
    }

    /**
     * Appends the interior points of a semicircular end cap around [center],
     * starting from [from] and bulging along the tangent direction.
     */
    private fun appendCapArc(
        contour: MutableList<WobblePoint>,
        center: WobblePoint,
        from: WobblePoint,
        halfWidth: Double,
        tangentX: Double,
        tangentY: Double,
    ) {
        val startAngle = atan2(from.y - center.y, from.x - center.x)
        val midAngle = startAngle + Math.PI / 2
        val direction = if (cos(midAngle) * tangentX + sin(midAngle) * tangentY >= 0) 1.0 else -1.0
        for (step in 1 until CAP_SEGMENTS) {
            val angle = startAngle + direction * (Math.PI * step / CAP_SEGMENTS)
            contour.add(WobblePoint(center.x + halfWidth * cos(angle), center.y + halfWidth * sin(angle)))
        }
    }

    /**
     * The displaced centerline mirrors the input subpath structure so a
     * future reveal mask's trim progresses across subpaths in the same order
     * as a stroke draw-on would. A one-point polyline stays one point — the
     * mask stroke's round cap is what will reveal the dot.
     */
    private fun displacedCenterline(
        polyline: WobblePolyline,
        params: WobbleParams,
        noise: SVGTurbulence,
    ): WobblePolyline = WobblePolyline(polyline.points.map { params.displace(it, noise) }, polyline.isClosed)
}

File: apps/ios/Pebbles/Features/Path/Render/Wobble/WobbleOutlineBuilder.swift
import CoreGraphics
import Foundation

/// The wobbled render of one artwork piece: the leaky filled ink, plus the
/// displaced centerline the appear animation's reveal mask strokes along.
final class WobbleArt {
    let ink: CGPath
    let centerline: CGPath

    init(ink: CGPath, centerline: CGPath) {
        self.ink = ink
        self.centerline = centerline
    }
}

/// Converts flattened centerline polylines into wobbled filled outlines,
/// porting the playground's baked-export algorithm: offset both stroke edges
/// from the centerline first, then displace every contour point independently.
/// The independent displacement is what makes the width breathe ("leaky") —
/// a stroked wobbled centerline would stay constant-width.
enum WobbleOutlineBuilder {

    /// End caps are semicircles of `capSegments` arcs (capSegments − 1
    /// interior points), matching the playground.
    private static let capSegments = 6
    /// A zero-length polyline (single-tap carve dot) becomes a circle with
    /// this many segments. The playground drops dots; we keep them visible,
    /// matching the current renderer's round-cap dot.
    private static let dotSegments = 12

    static func art(
        for polylines: [WobblePolyline],
        halfWidth: Double,
        params: WobbleParams,
        noise: SVGTurbulence
    ) -> WobbleArt {
        let ink = CGMutablePath()
        let centerline = CGMutablePath()
        for polyline in polylines {
            for contour in contours(for: polyline, halfWidth: halfWidth) {
                guard contour.count > 2 else { continue }
                appendClosedPolygon(contour.map { params.displace($0, using: noise) }, to: ink)
            }
            appendCenterline(for: polyline, params: params, noise: noise, to: centerline)
        }
        return WobbleArt(ink: ink.copy() ?? ink, centerline: centerline.copy() ?? centerline)
    }

    /// Un-displaced outline contours for one polyline. Internal so tests can
    /// assert point counts and winding without decoding a displaced path.
    static func contours(for polyline: WobblePolyline, halfWidth: Double) -> [[CGPoint]] {
        let points = polyline.points
        let count = points.count

        if count == 1 {
            let cx = Double(points[0].x)
            let cy = Double(points[0].y)
            let circle = (0..<dotSegments).map { i -> CGPoint in
                let angle = 2 * Double.pi * Double(i) / Double(dotSegments)
                return CGPoint(x: cx + halfWidth * cos(angle), y: cy + halfWidth * sin(angle))
            }
            return [circle]
        }

        let normals = Self.normals(for: points, closed: polyline.isClosed)
        let left = (0..<count).map { i in
            CGPoint(
                x: Double(points[i].x) + normals[i].x * halfWidth,
                y: Double(points[i].y) + normals[i].y * halfWidth
            )
        }
        let right = (0..<count).map { i in
            CGPoint(
                x: Double(points[i].x) - normals[i].x * halfWidth,
                y: Double(points[i].y) - normals[i].y * halfWidth
            )
        }

        if polyline.isClosed {
            // Outer ring + reversed inner ring: opposite windings make the
            // pair render as an annulus under nonzero fill.
            return [left, right.reversed()]
        }

        var contour = left
        appendCapArc(
            to: &contour,
            center: points[count - 1],
            from: left[count - 1],
            halfWidth: halfWidth,
            tangentX: Double(points[count - 1].x - points[count - 2].x),
            tangentY: Double(points[count - 1].y - points[count - 2].y)
        )
        contour.append(contentsOf: right.reversed())
        appendCapArc(
            to: &contour,
            center: points[0],
            from: right[0],
            halfWidth: halfWidth,
            tangentX: Double(points[0].x - points[1].x),
            tangentY: Double(points[0].y - points[1].y)
        )
        return [contour]
    }

    // MARK: - Pieces

    /// Per-point normals from neighbor tangents; endpoints use their single
    /// adjacent segment, closed rings wrap cyclically.
    private static func normals(for points: [CGPoint], closed: Bool) -> [SIMD2<Double>] {
        let count = points.count
        var result: [SIMD2<Double>] = []
        result.reserveCapacity(count)
        for i in 0..<count {
            let prev = closed ? points[(i - 1 + count) % count] : points[max(0, i - 1)]
            let next = closed ? points[(i + 1) % count] : points[min(count - 1, i + 1)]
            let tx = Double(next.x - prev.x)
            let ty = Double(next.y - prev.y)
            let rawLength = hypot(tx, ty)
            let length = rawLength == 0 ? 1 : rawLength   // playground: `|| 1`
            result.append(SIMD2(-ty / length, tx / length))
        }
        return result
    }

    /// Appends the interior points of a semicircular end cap around `center`,
    /// starting from `from` and bulging along the tangent direction.
    private static func appendCapArc(
        to contour: inout [CGPoint],
        center: CGPoint,
        from: CGPoint,
        halfWidth: Double,
        tangentX: Double,
        tangentY: Double
    ) {
        let cx = Double(center.x)
        let cy = Double(center.y)
        let startAngle = atan2(Double(from.y) - cy, Double(from.x) - cx)
        let midAngle = startAngle + .pi / 2
        let direction: Double = (cos(midAngle) * tangentX + sin(midAngle) * tangentY) >= 0 ? 1 : -1
        for step in 1..<capSegments {
            let angle = startAngle + direction * (.pi * Double(step) / Double(capSegments))
            contour.append(CGPoint(
                x: cx + halfWidth * cos(angle),
                y: cy + halfWidth * sin(angle)
            ))
        }
    }

    private static func appendClosedPolygon(_ points: [CGPoint], to path: CGMutablePath) {
        guard let first = points.first else { return }
        path.move(to: first)
        for point in points.dropFirst() {
            path.addLine(to: point)
        }
        path.closeSubpath()
    }

    /// The displaced centerline mirrors the input subpath structure so the
    /// reveal mask's `.trim` progresses across subpaths in the same order as
    /// today's stroke draw-on.
    private static func appendCenterline(
        for polyline: WobblePolyline,
        params: WobbleParams,
        noise: SVGTurbulence,
        to path: CGMutablePath
    ) {
        let displaced = polyline.points.map { params.displace($0, using: noise) }
        guard let first = displaced.first else { return }
        path.move(to: first)
        if displaced.count == 1 {
            // Zero-length segment: the mask stroke's round cap reveals the dot.
            path.addLine(to: first)
            return
        }
        for point in displaced.dropFirst() {
            path.addLine(to: point)
        }
        if polyline.isClosed {
            path.closeSubpath()
        }
    }
}
